Add timer and "no errors" reporting

ankritisingh opened the pull request:
1. Added a time which is displayed in seconds if less than a minute, and in minutes otherwise.

_**In seconds:**_

![image](https://github.com/user-attachments/assets/e1469843-2c63-4c25-b18b-cf8ca6e24f1f)

_**In minutes:**_

![image](https://github.com/user-attachments/assets/40ead8f3-dd1d-443a-a537-76b58e8c3810)

2. For stable results:
_It displays a line if the table is empty:_

![image](https://github.com/user-attachments/assets/08e2437a-d1da-4938-985f-d3d0f8045b76)

.
**However, if there were any changes reported in the "main.do" file, the line "no mismatches" does not appear when "stepping back into the file." I couldn't determine how to include this, but I believe it is fine (for now at least) since changes were reported at least once in that file. See below:**

![image](https://github.com/user-attachments/assets/6176a477-8830-4b47-aefe-fd5757cb53c7)

bbdaniels commented on src/ado/reprun.ado:
```diff
@@ -6,6 +6,9 @@ cap program drop   reprun
 
     version 14.1
+	
+    timer clear 1
+    timer on 1
```
I would suggest using a higher number -- if the user is using a timer, they may use `1` and this will cause an error or other collision.

bbdaniels commented on src/ado/reprun.ado:
```diff
@@ -146,6 +149,24 @@ qui {
+    timer off 		1
+    qui timer list 	1
+	
+    if `r(t1)' > 60 {
+      local duration :  di %9.3f `r(t1)'/60
+      noi di as res ""
+      noi di as res `"{phang}Total run time: `duration' minutes {p_end}"'
+    }
+	
+    else {
+      local duration : di %9.3f `r(t1)'
+      noi di as res ""
+      noi di as res `"{phang}Total run time: `duration' seconds {p_end}"'
+    }
+
```
`%9.2f` is probably sufficient, but you may instead consider reporting MM:SS when it runs over 60 seconds; and HH:MM:SS when it runs over 60 minutes, as this is more standard notation and easier for the user to interpret. Reporting days is probably unnecessary.

ankritisingh replied:
Updated it so now it is displayed as:

1. Seconds:
![Screenshot 2024-07-17 at 7 46 53 PM](https://github.com/user-attachments/assets/6e673adc-8c16-4bb2-8906-138267b441f2)

2. Minutes:
![Screenshot 2024-07-17 at 7 49 02 PM](https://github.com/user-attachments/assets/1d0f45f4-e761-4a70-a2a5-34399f3d5a85)

3. Hours:
![Screenshot 2024-07-17 at 7 46 21 PM](https://github.com/user-attachments/assets/bbec0eb6-2ba3-48b4-b947-105e15b6074c)

bbdaniels replied:
Nice!

bbdaniels commented on src/ado/reprun.ado:
```diff
@@ -625,6 +646,9 @@ end
     local prev_line2 ""
+	
+    * Local for empty tables
+    local any_lines_written 0	
```
You might be able to find a way to reset this when "stepping back"; alternatively, you might _not_ want to have this show up until each file is closed for the last time. This may or may not be predictable behavior.

I would suggest, since we will know what line the file has called a sub-do-file at, to report "no differences/errors in lines XX-YY" (don't worry about the exact message wording yet). I think we should be able to pick up XX/YY? @kbjarkefur?

ankritisingh replied:
I have not been able to figure this out. If there are no changes in the main file at all, then it does display "no mismatches...". But I have not been able to figure out how to reset it when "stepping back".  

Can think about this more after working on other things for the conference.

ankritisingh commented:
@bbdaniels Updated the help file [here](https://github.com/worldbank/repkit/blob/reprun-46-stable/src/mdhlp/reprun.md) for your review.

bbdaniels commented:
> @bbdaniels Updated the help file [here](https://github.com/worldbank/repkit/blob/reprun-46-stable/src/mdhlp/reprun.md) for your review.

Great start! The example should be modified a little bit. I would not include `bysort` here since we are still working on handling that functionality correctly -- just use `sort` and `gen` directly.

I would also split into two examples -- one where an issue is due to the lack of a seed, and one where it is due to an unstable sort. 

We should then include the exact output from `reprun` so the user can see it for all settings, and also see how early mismatches cause later ones. Then, for each of these two examples, show the resolution.

Finally, I would add one example of a recursive call to show that functionality.

ankritisingh commented:
> > @bbdaniels Updated the help file [here](https://github.com/worldbank/repkit/blob/reprun-46-stable/src/mdhlp/reprun.md) for your review.
> 
> Great start! The example should be modified a little bit. I would not include `bysort` here since we are still working on handling that functionality correctly -- just use `sort` and `gen` directly.
> 
> I would also split into two examples -- one where an issue is due to the lack of a seed, and one where it is due to an unstable sort.
> 
> We should then include the exact output from `reprun` so the user can see it for all settings, and also see how early mismatches cause later ones. Then, for each of these two examples, show the resolution.
> 
> Similarly, I would expand the example of a recursive call to show that functionality -- perhaps put a sorting issue in one file and a seeding issue in another, and show how solving them top-to-bottom works (and perhaps put in a cascading issue where solving bottom-to-top does not produce a solution).

Okay, thanks! I will update the examples.
I tried adding the results, but since we are using `ad_sthlp`, they are not being displayed properly in .md version and then won't display in the smcl version too. 
I will try again though.

ankritisingh commented:
> > > @bbdaniels Updated the help file [here](https://github.com/worldbank/repkit/blob/reprun-46-stable/src/mdhlp/reprun.md) for your review.
> > 
> > 
> > Great start! The example should be modified a little bit. I would not include `bysort` here since we are still working on handling that functionality correctly -- just use `sort` and `gen` directly.
> > I would also split into two examples -- one where an issue is due to the lack of a seed, and one where it is due to an unstable sort.
> > We should then include the exact output from `reprun` so the user can see it for all settings, and also see how early mismatches cause later ones. Then, for each of these two examples, show the resolution.
> > Similarly, I would expand the example of a recursive call to show that functionality -- perhaps put a sorting issue in one file and a seeding issue in another, and show how solving them top-to-bottom works (and perhaps put in a cascading issue where solving bottom-to-top does not produce a solution).
> 
> Okay, thanks! I will update the examples. I tried adding the results, but since we are using `ad_sthlp`, they are not being displayed properly in .md version and then won't display in the smcl version too. I will try again though.

@bbdaniels I updated the examples, splitting into two for clarity and reusing content for main.do to keep things concise. I tried to include the result, but the tables don't render well in Markdown or SMCL due to issues with merging cells. It seems that using `ad_sthlp` isn't a viable option here.

Maybe @kbjarkefur knows how to do it with `ad_sthlp`

Alternatively, we could link to a Dime Wiki article, as we've done for other commands. This would allow us to include screenshots and other details. I can work on this after the Stata conference.

bbdaniels commented:
Retargeting to dev so we can finish after conference
